###########################################
# saliva reference panel QC/preprocessing
# John Dou
###########################################


# Purpose: Preprocess and start QC of the saliva DNA methylation data from the EPIC array
# 
# Inputs: Prjt_321_Epicore_20190511_CP2912.csv - details about sequenced samples
#         idats -                                generated by EPIC
#         pd.qc.rds -                            full demographics by meth_id
#         ewatools-meth.rds -                    list of EPIC data
#
# Outputs: RGset, plots showing QC data


#======================================== load in data ========================================#

library(ewastools)
library(minfi)

#construct file names
samp.sheet <- read.csv("/nfs/turbo/bakulski1/Datasets/Saliva/Bakulsky_EPIC01_KB-199/raw_data/SampleSheet/Prjt_321_Epicore_20190511_CP2912.csv", skip=7, header=T)
files <- paste0(samp.sheet$Sentrix_ID,'/',samp.sheet$Sentrix_ID,'_',samp.sheet$Sentrix_Position)

#read in idats in ewastools and minfi
setwd("/nfs/turbo/bakulski1/Datasets/Saliva/Bakulsky_EPIC01_KB-199/raw_data/ImageData/")
meth <- read_idats(files)
RGset <- read.metharray(basenames=files,verbose=T)

setwd("/nfs/turbo/bakulski1/People/johndou/Saliva_Ref_Panel/01_Raw/")
saveRDS(meth,file="ewastools-meth.rds")
saveRDS(RGset,file="RGset.rds")

rm(samp.sheet,files)

#======================================== checks with ewastools ========================================#

library(ewastools)
library(magrittr)


### start constructing phenotype file, merge in demog info
# this section is not relevant to GEO file;
pd <- read.csv("/nfs/turbo/bakulski1/Datasets/Saliva/Bakulsky_EPIC01_KB-199/raw_data/SampleSheet/Prjt_321_Epicore_20190511_CP2912.csv", skip=7, header=T)
demog <- read.csv("/nfs/turbo/bakulski1/People/johndou/Saliva_Ref_Panel/05-28-19 demog_full.csv")

#Lauren's:
pd.qc <- readRDS("pd.qc.rds")
pd <- read.csv("C:/Users/HP/Documents/Research 2019/Prjt_321_Epicore_20190511_CP2912.csv", skip=7, header=T)
demog <- read.csv("C:/Users/HP/Documents/Research 2019/05-28-19 demog_full.csv")

pd <- pd[,1:9]
pd$meth_id <- paste0(pd$Sentrix_ID,'_',pd$Sentrix_Position) #matching up sample names with IDs
pd$celltype <- gsub('Dupl','',pd$Sample_Group) %>% gsub('.*_','',.)
pd$Sample_Group <- gsub('_.*','',pd$Sample_Group)
demog$Sample_Group <- gsub('_.*','',demog$Sample.Name)
demog <- demog[,-c(1,2)] %>% unique()

pd.demog <- merge(pd, demog, by = "Sample_Group", all.x=T, all.y=F)

rm(pd,demog)

### control probe metrics

setwd("/nfs/turbo/bakulski1/People/johndou/Saliva_Ref_Panel/02_QC_checks/")
meth <- readRDS("/nfs/turbo/bakulski1/People/johndou/Saliva_Ref_Panel/01_Raw/ewatools-meth.rds")
#Lauren's
meth <- readRDS("C:/Users/HP/Documents/Research 2019/06-10-19 saliva ewatools-meth.rds")

ctrl.metrics <- control_metrics(meth)


#pdf('control_metrics.pdf')
pdf("07-15-19 saliva control_metrics.pdf")
for(metric in names(ctrl.metrics)){
  stripchart(ctrl.metrics[[metric]],method="jitter",pch=4,xlab=metric)
  abline(v=attr(ctrl.metrics[[metric]],'threshold'),col=2,lty=3)
}
dev.off()

#flag for any metric threshold fail
fail.flag <- sample_failure(ctrl.metrics)
table(fail.flag) #false = 72, true = 8
pd.flag <- pd.demog
pd.flag <- pd.flag[match(meth$meta$sample_id,pd.flag$meth_id),]
pd.flag$control_probe_flag <- fail.flag

#flags for specific metric threshold fail
failed = sapply(ctrl.metrics,function(metric){
  metric < attr(metric,"threshold") 
})

colnames(failed) %<>% gsub(' |/|-','_',.)

pd.flag <- cbind(pd.flag,failed)

#counts of flagged samples
failed %<>% data.frame()
failed$Any <- fail.flag
counts <- colSums(failed)
#write.csv(counts,file='control.flag.counts.csv')
write.csv(counts, file = "07-15-19 control.flag.counts.csv")

#which samples flagged
samp.flag <- pd.flag[pd.flag$control_probe_flag,c('Sample.ID',colnames(failed)[-18])]
samp.flag <- t(samp.flag) #makes a dataset of each control metric by sample fraction
#write.csv(samp.flag,file='control_probe_flagged_samples.csv')
write.csv(samp.flag, file = "07-15-19 control_probe_flagged_samples.csv")

rm(ctrl.metrics,fail.flag,failed,metric,counts,samp.flag)


### sex checks

#correct dye bias first
meth %<>% correct_dye_bias()

#check pd file lines up with meth object again
rownames(pd.flag) <- pd.flag$meth_id
identical(rownames(pd.flag),meth$meta$sample_id)

#X and Y normalized intensities
pd.sex <- pd.flag
pd.sex[,c('X','Y')] <- check_sex(meth)

#predicted sex
pd.sex$Male_label <- ifelse(pd.sex$Male_label=='Female','f','m')
pd.sex$predicted_sex <- predict_sex(X=pd.sex$X, Y=pd.sex$Y, male=which(pd.sex$Male_label=='m'), female=which(pd.sex$Male_label=='f'))

table(pd.sex$predicted_sex,pd.sex$Male_label)

pd.sex[is.na(pd.sex$predicted_sex) | pd.sex$Male_label != pd.sex$predicted_sex,c('Sample.ID','Male','predicted_sex')]

pdf('sex_plot_ewastools2.pdf')
  tch <- ifelse(is.na(pd.sex$predicted_sex),3,
                ifelse(pd.sex$predicted_sex=='f',1,4))
  plot(Y ~ X,data=pd.sex,pch=tch,asp=1,xlab="Normalized X chromosome intensities",ylab="Normalized Y chromosome intensities")
  diff <- pd.sex[is.na(pd.sex$predicted_sex) | pd.sex$Male_label!=pd.sex$predicted_sex,]
  tch.diff <- ifelse(is.na(diff$predicted_sex),3,
                     ifelse(diff$predicted_sex=='f',1,4))
  points(Y ~ X,data=diff,pch=tch.diff,col=2)
  legend("topright",pch=c(1,4,3),legend=c("female","male","unassigned"))
dev.off()

saveRDS(pd.sex,file='pd.sex.rds')
rm(diff,tch,tch.diff)

### detection p 
detP <- detectionP(meth)

ewastools.detP <- detP$detP
samps <- detP$meta$sample_id
probes <- detP$manifest$probe_id
colnames(ewastools.detP) <- samps
rownames(ewastools.detP) <- probes

save(ewastools.detP,file='ewastools.detP.rda')
load('ewastools.detP.rda')

pdf('detP_eval_ewastools.pdf')
  eval_detP_cutoffs(detP,males=which(pd.sex$Male_label=='m'),females=which(pd.sex$Male_label=='f'))
  eval_detP_cutoffs(detP,males=which(pd.sex$predicted_sex=='m'),females=which(pd.sex$predicted_sex=='f'))
dev.off()

failedP <- ewastools.detP>0.05
per.samp<-colMeans(failedP,na.rm=T) # Fraction of failed positions per sample
summary(per.samp)
per.probe<-rowMeans(failedP,na.rm=T) # Fraction of failed samples per position
summary(per.probe)
sum(per.samp>0.1) #How many samples had more than 10% of sites fail?
# 17
sum(per.probe>0.05,na.rm=T) # How many positions failed >5% samples
# 179,136

pdf('detP-hist.pdf')
  hist(per.samp,breaks=20,xlab='Fraction of Failed Positions Per Sample',main='All Samples')
  hist(per.samp[per.samp<0.2],breaks=40,xlab='Fraction of Failed Positions Per Sample',main='Samples with <20% Failed Probes')
  hist(per.probe,breaks=20,xlab='Fraction of Failed Samples Per Position',main='All Probes')
  hist(per.probe[per.probe<0.3],breaks=60,xlab='Fraction of Failed Samples Per Position',main='Probes with <30% Sampled Failed')
dev.off()

save(per.probe,file='ewastools_per_probe_fail_rate.rds')

pd.detp <- pd.sex
pd.detp$ew_probe_fail_pct <- per.samp[rownames(pd.detp)]


save(pd.detp,file='pd.detp.rds')
rm(ewastools.detP,failedP,per.probe,per.samp,probe.fail,probes,samps)


### genotype calling

beta <- dont_normalize(meth)
snps <- meth$manifest[probe_type=='rs',index]
snps <- beta[snps,]
genotypes <- call_genotypes(snps,learn=FALSE)

# saveRDS(genotypes,file='called-genotypes.rds')
# genotypes <- readRDS('called-genotypes.rds')

#double check ordering
identical(colnames(genotypes$snps),rownames(pd.detp))

#check agreement of samples from same person based on snps
check.snp <- check_snp_agreement(genotypes,donor_ids=pd.detp$Sample_Group,sample_ids=pd.detp$Sample.ID)

pdf('snp_agreement.pdf')
  ewastools:::agreement_(genotypes,pd.detp$Sample_Group,pd.detp$Sample.ID,names=c('Different Donor','Same Donor'))
dev.off()

#agreement matrix
donor_ids <- pd.detp$Sample_Group
sample_ids <- pd.detp$Sample.ID

library(data.table)
J = ncol(genotypes$snps)
weights = 1-genotypes$outliers
gamma = genotypes$gamma

stopifnot(anyDuplicated(sample_ids)==0)
stopifnot(length(sample_ids) == J)

sample_ids = as.character(sample_ids)

conflicts = cbind(CJ(donor_ids,donor_ids,sorted=FALSE),CJ(sample_ids,sample_ids,sorted=FALSE))
setcolorder(conflicts,c(1,3,2,4))
setnames(conflicts,1:4,c('donor1','sample1','donor2','sample2'))

d = matrix(NA_real_,nrow=J,ncol=J)
for(j in 1:J){
  tmp = 
    (weights * gamma[[1]]) * (weights[,j] * gamma[[1]][,j]) +
    (weights * gamma[[2]]) * (weights[,j] * gamma[[2]][,j]) +
    (weights * gamma[[3]]) * (weights[,j] * gamma[[3]][,j])
  
  d[,j] = colSums(tmp,na.rm=TRUE) / colSums(weights*weights[,j],na.rm=TRUE)
  
}

# first drop the duplicate entries and the diagonal
d[upper.tri(d,diag=TRUE)] = NA_real_
conflicts$agreement = as.numeric(d)
rm(tmp,d)
conflicts = conflicts[!is.na(agreement)]

saveRDS(conflicts,file='full_samp_agreements.csv')

#make excel file of the unexpected relations
library(openxlsx)
wb <- createWorkbook()
for(i in 1:length(check.snp)){
  addWorksheet(wb,i)
  writeData(wb,sheet=i,x=check.snp[[i]])
  setColWidths(wb,sheet=i,cols=1:5,widths='auto')
}
saveWorkbook(wb,"snp_innapropriate_sample_relations.xlsx",overwrite=T)

#logodd of being outlier on snp probes
pd.snp <- pd.detp
pd.snp$snp_logodd_outlier = snp_outliers(genotypes)
table(pd.snp$snp_logodd_outlier > -4)
# FALSE  TRUE
# 63    17


saveRDS(pd.snp,file='pd.snp.rds')

rm(beta,meth,genotypes,snps,i,pd.demog,pd.detp,pd.flag,pd.sex,check.snp,pd.snp,wb)



#======================================== checks with minfi ========================================#

library(minfi)
library(magrittr)

setwd("/nfs/turbo/bakulski1/People/johndou/Saliva_Ref_Panel/02_QC_checks/")
pd <- readRDS('pd.snp.rds')
RGset <- readRDS("/nfs/turbo/bakulski1/People/johndou/Saliva_Ref_Panel/01_Raw/RGset.rds")


### Overall intensity: M vs. U

rawMSet <- preprocessRaw(RGset)
rawMSet
#M signal per probe, per sample
Meth<-getMeth(rawMSet)
Meth[1:5,1:5]
#M signal per probe, per sample
Unmeth<-getUnmeth(rawMSet)
Unmeth[1:5,1:5]

MQC<-log2(colMedians(Meth))
UQC<-log2(colMedians(Unmeth))

intensities <- data.frame(meth_id=colnames(Meth),MQC=MQC,UQC=UQC)

#append intensities to pd 
pd.int <- merge(pd,intensities,by='meth_id',all.x=T)
dim(pd.int)
pd <- pd.int

#plot M vs U
library(RColorBrewer)
myColors <- brewer.pal(12,'Set3')

slide <- as.factor(pd$Sentrix_ID)

pdf("MvsUplot-group.pdf")
palette(myColors)
plot(pd$UQC, pd$MQC, col=slide, main="M vs. U QC by Sentrix ID", pch=16, xlab="Log2 Median Unmethylated Intensity", ylab="Log2 Median Methylated Intensity", cex.lab=1.2, cex.main=2)
legend("topleft",levels(slide), fill=myColors)
dev.off()

table(pd$UQC<9.5 & pd$MQC<9.5)
pd$lowest_int <- pd$UQC<9.5 & pd$MQC<9.5

pdf("MvsUplot-labeled.pdf")
palette(myColors)
plot(pd$UQC, pd$MQC, col=slide, main="M vs. U QC by Sentrix ID", pch=16, xlab="Log2 Median Unmethylated Intensity", ylab="Log2 Median Methylated Intensity", cex.lab=1.2, cex.main=2)
legend("topleft",levels(slide), fill=myColors)
with(pd[pd$UQC<9.5 & pd$MQC<9.5,],
     text(UQC, MQC, labels=Sample.ID, pos=4, cex=0.6))
dev.off()

#Raw density plot
beta.raw<-getBeta(rawMSet)

pdf("Density-plot-Raw.pdf")
  densityPlot(beta.raw, sampGroups=pd$celltype, main="Raw Beta by Cell Type")
  densityPlot(beta.raw, sampGroups=pd$lowest_int, main="Raw Beta by UQC and MQC < 9.5")
dev.off()

rm(beta.raw,intensities,Meth,MQC,Unmeth,UQC,slide,rawMSet,pd.int)


### detection P
detP <- detectionP(RGset)
dim(detP)
detP[1:5, 1:5]
# save(detP,file="minfi-detP.rda")
# load("minfi-detP.rda")
failedP <- detP>0.01
per.samp<-colMeans(failedP) # Fraction of failed positions per sample
summary(per.samp)
per.probe<-rowMeans(failedP) # Fraction of failed samples per position
summary(per.probe)
sum(per.samp>0.05) #How many samples had more than 5% of sites fail?
# 2
sum(per.probe>0.05) # How many positions failed in at least 5% of samples? 
# 17095
sum(per.probe>0.02) # How many positions failed in at least 2 samples? 
# 80713

write.csv(pd[pd$mf_probe_fail_pct>0.01,c('Sample.ID','mf_probe_fail_pct')],file='minfi_detp_failGT1pct.csv')
save(per.probe,file='minfi_per_probe_fail_rate.rds')

pdf('detP-hist-minfi.pdf')
  hist(per.samp,breaks=20,xlab='Fraction of Failed Positions Per Sample',main='All Samples')
  hist(per.samp[per.samp<0.2],breaks=40,xlab='Fraction of Failed Positions Per Sample',main='Samples with <20% Failed Probes')
  hist(per.probe,breaks=20,xlab='Fraction of Failed Samples Per Position',main='All Probes')
  hist(per.probe[per.probe<0.15],xlab='Fraction of Failed Samples Per Position',main='Probes with <15% Samples Failed')
dev.off()

#put detp fail rate info into the pd data frame
pd$mf_probe_fail_pct <- per.samp[pd$meth_id]
pd[,c('Sample.ID','ew_probe_fail_pct','mf_probe_fail_pct')]
#saveRDS(pd,file='pd.qc.rds')


### NOOB
noob <- preprocessNoob(RGset, offset=15, dyeCorr=TRUE, verbose = TRUE)
rm(RGset)
# saveRDS(noob,file="noob.rds")
# noob <- readRDS('noob.rds')
beta <- getBeta(noob)
pd <- readRDS('pd.qc.rds')


rownames(pd) <- pd$meth_id
table(rownames(pd) %in% colnames(beta))
pd <- pd[colnames(beta),]
identical(colnames(beta),rownames(pd))

pdf("Density-plot-noob-k.pdf")
  densityPlot(beta, sampGroups=pd$snp_logodd_outlier>-4, main="Log Odd SNP Probe Outlier > -4")
  pd.sal02out <- pd[pd$Sample.ID != 'SAL02_CD45pos',]
  densityPlot(beta[,rownames(pd.sal02out)], sampGroups=pd.sal02out$snp_logodd_outlier>-4, main="Log Odd SNP Probe Outlier > -4")
  densityPlot(beta[,rownames(pd.sal02out)], sampGroups=pd.sal02out$lowest_int, main="Noob Beta by Low Intensity")
  densityPlot(beta[,rownames(pd.sal02out)], sampGroups=pd.sal02out$mf_probe_fail_pct>0.01, main="Noob Beta by DetP Fail > 1% Probes")
dev.off()


### PCA
prin <- prcomp(t(beta),center=T,scale.=F)
#save(prin,file='pcs-beta.rda')
load('pcs-beta.rda')
pd <- readRDS('pd.qc.rds')
rownames(pd) <- pd$meth_id

library(RColorBrewer)
graphColors <- brewer.pal(8,'Set1')
graphColors2 <- c("#bf1b08",
                  "#0281e9",
                  "#64ba22",
                  "#cb4cd0",
                  "#b1d34e",
                  "#503db8",
                  "#dea800",
                  "#f59bff",
                  "#00bb6f",
                  "#a50347",
                  "#64dbae",
                  "#ff8edf",
                  "#859300",
                  "#86baff",
                  "#ff8367",
                  "#235e34",
                  "#ff94ae",
                  "#79461f",
                  "#69447e",
                  "#edbd8c",
                  "#b17291")


prin <- prcomp(t(beta[,rownames(pd.sal02out)]),center=T,scale.=F)
#save("pcs-beta-no_sal02cd45pos.rda")
load("pcs-beta-no_sal02cd45pos.rda")

identical(rownames(prin$x),rownames(pd))
pd <- pd[rownames(prin$x),]
identical(rownames(prin$x),rownames(pd))

pdf('PCplot_person.pdf')
palette(graphColors2)
pairs(prin$x[,1:4], col=as.factor(pd$Sample_Group), labels=c("PC1", "PC2", "PC3", "PC4"), pch=1, cex=1.1, oma=c(2,2,2,13))
par(xpd=T)
legend(0.85,0.9,legend=levels(as.factor(pd$Sample_Group)),fill=as.factor(pd$Sample_Group)%>%as.numeric%>%as.factor%>%levels, title="Person")
dev.off()

pdf('PCplot_sex.pdf')
palette(graphColors)
pairs(prin$x[,1:4], col=as.factor(pd$predicted_sex), labels=c("PC1", "PC2", "PC3", "PC4"), pch=1, cex=1.1, oma=c(2,2,2,13))
par(xpd=T)
legend(0.85,0.9,legend=levels(as.factor(pd$predicted_sex)),fill=as.factor(pd$predicted_sex)%>%as.numeric%>%as.factor%>%levels, title="Sex")
dev.off()

pdf('PCplot_celltype.pdf')
palette(graphColors)
pairs(prin$x[,1:4], col=as.factor(pd$celltype), labels=c("PC1", "PC2", "PC3", "PC4"), pch=1, cex=1.1, oma=c(2,2,2,13))
par(xpd=T)
legend(0.85,0.9,legend=levels(as.factor(pd$celltype)), fill=as.factor(pd$celltype)%>%as.numeric%>%as.factor%>%levels, title="Cell Type")
dev.off()

pdf('PC_names.pdf')
palette(graphColors)
plot(prin$x[,1], prin$x[,2], col=as.factor(pd$celltype), xlab='PC1',ylab='PC2', pch=2, cex=0.5, oma=c(2,2,2,13))
text(prin$x[,1], prin$x[,2], col=graphColors[as.factor(pd$celltype)], labels=pd$Sample.ID, cex=0.3)
legend('topleft',legend=levels(as.factor(pd$celltype)), fill=as.factor(pd$celltype)%>%as.numeric%>%as.factor%>%levels, title="Cell Type")

plot(prin$x[,1], prin$x[,3], col=as.factor(pd$celltype), xlab='PC1',ylab='PC3', pch=2, cex=0.5, oma=c(2,2,2,13))
text(prin$x[,1], prin$x[,3], col=graphColors[as.factor(pd$celltype)], labels=pd$Sample.ID, cex=0.3)

plot(prin$x[,1], prin$x[,4], col=as.factor(pd$celltype), xlab='PC1',ylab='PC4', pch=2, cex=0.5, oma=c(2,2,2,13))
text(prin$x[,1], prin$x[,4], col=graphColors[as.factor(pd$celltype)], labels=pd$Sample.ID, cex=0.3)

plot(prin$x[,2], prin$x[,3], col=as.factor(pd$celltype), xlab='PC2',ylab='PC3', pch=2, cex=0.5, oma=c(2,2,2,13))
text(prin$x[,2], prin$x[,3], col=graphColors[as.factor(pd$celltype)], labels=pd$Sample.ID, cex=0.3)

plot(prin$x[,2], prin$x[,4], col=as.factor(pd$celltype), xlab='PC2',ylab='PC4', pch=2, cex=0.5, oma=c(2,2,2,13))
text(prin$x[,2], prin$x[,4], col=graphColors[as.factor(pd$celltype)], labels=pd$Sample.ID, cex=0.3)

plot(prin$x[,3], prin$x[,4], col=as.factor(pd$celltype), xlab='PC3',ylab='PC4', pch=2, cex=0.5, oma=c(2,2,2,13))
text(prin$x[,3], prin$x[,4], col=graphColors[as.factor(pd$celltype)], labels=pd$Sample.ID, cex=0.3)

dev.off()

png('PC_names.png',height=12,width=12,units='in',res=300)
palette(graphColors)
plot(prin$x[,1], prin$x[,2], col=as.factor(pd$celltype), xlab='PC1',ylab='PC2', pch=2, cex=1.1)
text(prin$x[,1], prin$x[,2], col=graphColors[as.factor(pd$celltype)], labels=pd$Sample.ID, cex=0.8)
legend('topleft',legend=levels(as.factor(pd$celltype)), fill=as.factor(pd$celltype)%>%as.numeric%>%as.factor%>%levels, title="Cell Type")
dev.off()

#highlight dups
dups <- pd[pd$NOTE=='technical duplicate','Sample.ID'] %>% as.character
dups <- c(dups,gsub('Dupl','',dups))
pd.dup <- pd[pd$Sample.ID %in% dups,]
prin.dup <- prin$x[rownames(prin$x) %in% pd.dup$meth_id,]
prin.dup <- prin.dup[rownames(pd.dup),]

pdf('PC_names_dup.pdf')
palette(graphColors)
plot(prin$x[,1], prin$x[,2], col=as.factor(pd$celltype), xlab='PC1',ylab='PC2', pch=2, cex=0.5, oma=c(2,2,2,13))
text(prin.dup[,1], prin.dup[,2], col='black', labels=pd.dup$Sample.ID, cex=0.6)
dev.off()

center_rowmeans <- function(beta)
  {
  beta.center <- rowMeans(beta)
  sweep(beta, 1, beta.center, '-')
  }

centered.beta <- center_rowmeans(beta)

donors <- pd.dup$Sample_Group %>% unique()

pdf('corr_dups_plus_bonus.pdf')
for(person in donors)
  {
  what.samps <- pd.dup[pd.dup$Sample_Group==person,'Sample.ID']
  beta.comp <- centered.beta[,pd.dup[pd.dup$Sample_Group==person,'meth_id']]
  smoothScatter(beta.comp[,1],beta.comp[,2],main=paste0(what.samps[1],' and ', what.samps[2], ' Pearson r=',round(cor(beta.comp[,1],beta.comp[,2]),2)))
  }

beta.comp <- centered.beta[,pd[pd$Sample.ID %in% c('SAL22_largeDupl','SAL18_whole'),'meth_id']]
smoothScatter(beta.comp[,1],beta.comp[,2],main=paste0('SAL22_largeDupl and SAL18_whole Pearson r=',round(cor(beta.comp[,1],beta.comp[,2]),2)))

beta.comp <- centered.beta[,pd[pd$Sample.ID %in% c('SAL22_largeDupl','SAL18_large'),'meth_id']]
smoothScatter(beta.comp[,1],beta.comp[,2],main=paste0('SAL22_largeDupl and SAL18_large Pearson r=',round(cor(beta.comp[,1],beta.comp[,2]),2)))

beta.comp <- centered.beta[,pd[pd$Sample.ID %in% c('SAL18_large','SAL18_whole'),'meth_id']]
smoothScatter(beta.comp[,1],beta.comp[,2],main=paste0('SAL18_large and SAL18_whole Pearson r=',round(cor(beta.comp[,1],beta.comp[,2]),2)))

beta.comp <- centered.beta[,pd[pd$Sample.ID %in% c('SAL10_CD45pos','SAL10_large'),'meth_id']]
smoothScatter(beta.comp[,1],beta.comp[,2],main=paste0('SAL10_CD45pos and SAL10_large Pearson r=',round(cor(beta.comp[,1],beta.comp[,2]),2)))

beta.comp <- centered.beta[,pd[pd$Sample.ID %in% c('SAL01_CD45pos','SAL01_CD45neg'),'meth_id']]
smoothScatter(beta.comp[,1],beta.comp[,2],main=paste0('SAL01_CD45pos and SAL01_CD45neg Pearson r=',round(cor(beta.comp[,1],beta.comp[,2]),2)))

beta.comp <- centered.beta[,pd[pd$Sample.ID %in% c('SAL15_large','SAL15_whole'),'meth_id']]
smoothScatter(beta.comp[,1],beta.comp[,2],main=paste0('SAL15_large and SAL15_whole Pearson r=',round(cor(beta.comp[,1],beta.comp[,2]),2)))

beta.comp <- centered.beta[,pd[pd$Sample.ID %in% c('SAL14_large','SAL14_whole'),'meth_id']]
smoothScatter(beta.comp[,1],beta.comp[,2],main=paste0('SAL14_large and SAL14_whole Pearson r=',round(cor(beta.comp[,1],beta.comp[,2]),2)))


dev.off()


### info about questionable samples
library(magrittr)
library(minfi)

setwd("/nfs/turbo/bakulski1/People/johndou/Saliva_Ref_Panel/02_QC_checks/")
pd <- readRDS('pd.qc.rds')

pd.q <- pd[pd$Sample.ID %in% c('SAL01_CD45pos','SAL17_large','SAL21_large'),]
pd.q <- pd.q[,c('Sample.ID','ew_probe_fail_pct','mf_probe_fail_pct','snp_logodd_outlier','lowest_int','control_probe_flag')]

noob <- readRDS('noob.rds')
beta <- getBeta(noob)
rm(noob)

rownames(pd) <- pd$meth_id
table(rownames(pd) %in% colnames(beta))
pd <- pd[colnames(beta),]
identical(colnames(beta),rownames(pd))

pd.red <- pd[!(pd$Sample.ID %in% c('SAL02_CD45pos','SAL22_largeDupl','SAL21_whole')),]
beta.red <- beta[,rownames(pd.red)]
look.at.these <- ifelse(pd.red$Sample.ID %in% c('SAL01_CD45pos','SAL17_large','SAL21_large'),as.character(pd.red$Sample.ID),'Other Sample')

pdf("Density-plot-investigate.pdf")
densityPlot(beta.red, sampGroups=look.at.these, main="")
dev.off()

load("pcs-beta-no_sal02cd45pos.rda")

pd.part <- pd.red[pd.red$celltype != 'whole' & pd.red$celltype != 'oragene',]
prin$x <- prin$x[rownames(pd.part),]

library(RColorBrewer)
graphColors <- brewer.pal(8,'Set1')

pdf('PCplot_celltype_nobulk.pdf')
palette(graphColors)
pairs(prin$x[,1:4], col=as.factor(pd.part$celltype), labels=c("PC1", "PC2", "PC3", "PC4"), pch=1, cex=1.1, oma=c(2,2,2,13))
par(xpd=T)
legend(0.85,0.9,legend=levels(as.factor(pd.part$celltype)), fill=as.factor(pd.part$celltype)%>%as.numeric%>%as.factor%>%levels, title="Cell Type")
dev.off()


conflicts <- readRDS('full_samp_agreements.rds')
con17 <- conflicts[conflicts$sample1=='SAL17_large' | conflicts$sample2=='SAL17_large',]
con17 <- con17[order(-con17$agreement),]
write.csv(con17[1:10,],'sal17_large_agree.csv')
