---
title: "Dropping Probes and Samples"
author: "Lauren Middleton"
date: "6/11/2019"
output: html_document
---
```{r}
# Purpose: QC the saliva beta matrix and drop samples and probes
# 
# Inputs: noob_beta.rds -   background-corrected dataset
#         pd.qc.rds -       full demographics by meth_id
#         minfi-detP.rda  - generated by code file 2.0 saliva_preprocess, % samples that failed detection p per probe
#         cross_probes_info_EPIC_Pidsley_2016.csv - list of the cross-reactive probes as generated by Pidsley et al. 2016
#
# Outputs:  beta_final_all_samps: "06-11-19 beta_final_all_samps.rds" - beta matrix of all samples including whole and oragene
#           pd_final_all_samps:   "06-11-19 pd_final_all_samps.rds"   - details of all samples including whole and oragene
#           pd_final:             "06-11-19 pd_final.rds"             - details of sample fractions (epithelial and immune)
#           beta_final:           "06-11-19 beta_final.rds"           - beta matrix of sample fractions (epithelial and immune)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir="C:/Users/HP/Documents/Research 2019")
knitr::opts_chunk$set(root.dir="C:/Users/HP/Documents/Research 2019")
```


## Install packages and load libraries;
```{r, include = FALSE}
# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("IlluminaHumanMethylationEPICanno.ilm10b2.hg19")

# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("GenomeInfoDb")

# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("bumphunter")

# install.packages("digest")
# install.packages("RSQLite")
# install.packages("bit")
# install.packages("XML")
# install.packages("quadprog")
# install.packages("data.table")
library(bumphunter)
library(minfi)
```

## Load illumina annotation library;
```{r, include = FALSE}
library(IlluminaHumanMethylationEPICanno.ilm10b2.hg19)
```


```{r}
data(Locations) #load locations data from illumina;

head(Locations) #X and Y chromosomes are called "chrX" or "chrY";
```

# Set up and read in datasets;
```{r}
setwd("C:/Users/HP/Documents/Research 2019")

#read in background corrected dataset from John, check dimensions;
noob_beta <- readRDS("noob_beta.rds")
dim(noob_beta) #866091 probes    80 samples;

#read in demographics and sample QC dataset from John, check to make sure it read in correctly;
pd.qc <- readRDS("pd.qc.rds")
head(pd.qc)

#Load probe fail dataset from John (from 2.0 saliva_preprocess)
#this is called per.probe (recalculated below)
#gives % samples that failed detection p per probe
#this is saved as detP in R
load("minfi-detP.rda")
dim(detP)
detP[1:5, 1:5] #this is p-values of whether or not the probe is detected (p<0.01);
length(detP) #69287280

#read in dataset of cross-reactive probes from Pidsley paper;
cross_probes <- read.csv("cross_probes_info_EPIC_Pidsley_2016.csv")
head(cross_probes)
```

# Check to see how many samples have >3% probes fail;
```{r}
head(pd.qc) 
table(pd.qc$mf_probe_fail_pct > 0.03) #3 samples have >3%;

#which 3 samples are these
w <- which(pd.qc$mf_probe_fail_pct > 0.03)
w
#sample 6 = 21_whole, 58 = 21_large, 59 = 02_CD45pos

table(pd.qc$NOTE) #how many technical duplicates: 6;
```

# First drop the duplicate samples;
```{r}
#check dimensions;
dim(pd.qc)

#subset the duplicate IDs
dups <- pd.qc[pd.qc$NOTE == "technical duplicate", "Sample.ID"]

#take out the Dupl ending to get the original endings
orig <- gsub("Dupl", "", dups)

#subset original pd to get originals and duplicates
pd_dups <- pd.qc[pd.qc$Sample.ID %in% orig | pd.qc$Sample.ID %in% dups,]
#put the same samples next to each other
pd_dups <- pd_dups[order(pd_dups$Sample.ID), c("Sample.ID",
                                               "Male_label",
                                               "predicted_sex",
                                               "control_probe_flag",
                                               "mf_probe_fail_pct",
                                               "MQC",
                                               "UQC")]

#better samples (better ones have lower fail rates or actually agree with their predicted sex)
drop_this <- c("SAL03_large", "SAL05_largeDupl", "SAL14_large", "SAL15_largeDupl", "SAL18_largeDupl", "SAL22_largeDupl")

#keep everything that is not the worse samples;
pd_dropdups <- pd.qc[!pd.qc$Sample.ID %in% drop_this, ]
#check dropped dimensions;
dim(pd_dropdups) #74 64
```


# Side quest: (4/20/20) calculate the mean centered correlation between the beta values of the duplicates
```{r}
#can't use beta_final_all_samps because we removed the duplicates already so I have to redo the entire preprocess
#Here we go
#if you want notes about each of these steps, see other parts of the code

#load things
library(IlluminaHumanMethylationEPICanno.ilm10b2.hg19)
data(Locations) #load locations data from illumina;
head(Locations) #X and Y chromosomes are called "chrX" or "chrY";
noob_beta <- readRDS("noob_beta.rds")
load("minfi-detP.rda")
cross_probes <- read.csv("cross_probes_info_EPIC_Pidsley_2016.csv")
pd.qc <- readRDS("pd.qc.rds")

dim(pd.qc) #80 64
table(pd.qc$celltype) # 9 cd45neg fractions
cd45neg_samps <- pd.qc$celltype == "CD45neg"
table(cd45neg_samps) #9 TRUE
pd_final_saliva1 <-  pd.qc[!cd45neg_samps, ]
dim(pd_final_saliva1) #71 64 #removed all the cd45negs

pd_final_saliva2 <- pd_final_saliva1[pd_final_saliva1$Sample.ID != "SAL19_large", ]
dim(pd_final_saliva2) #70 64
pd_final_saliva <- pd_final_saliva2

pd_dropagree <- pd_final_saliva[pd_final_saliva$Sample.ID != "SAL17_large", ]
dim(pd_dropagree) #69 64

low_pdrop <- pd_dropagree[pd_dropagree$mf_probe_fail_pct <0.03, ]
failedP <- detP > 0.01
failedP <- failedP[ ,low_pdrop$meth_id]
per.probe <- rowMeans(failedP)
table(per.probe >0.05)
#  FALSE   TRUE 
# 855800  10291
rm(detP, failedP)
detp_fail <- per.probe[per.probe >0.05]
length(detp_fail) #10291
detp_fail <- names(detp_fail)
beta_nodetp <- noob_beta[!rownames(noob_beta) %in% detp_fail, ]
dim(beta_nodetp) #855800     80

beta_nocross <- beta_nodetp[!rownames(beta_nodetp) %in% cross_probes$Probe, ]
sex_probes <- Locations[Locations$chr == "chrX" | Locations$chr == "chrY", ]
beta_noXY <- beta_nocross[!rownames(beta_nocross) %in% rownames(sex_probes), ]
pd_final <- low_pdrop
beta_final <- beta_noXY[ , pd_final$meth_id]
table(pd_final$celltype)
# CD45pos   large oragene   whole 
#      20      24       4      18
#18+6 large
dim(beta_final) #794360     66
rm(beta_nocross, beta_nodetp, beta_noXY, cross_probes, noob_beta, Locations)

#save these datasets to avoid having to rerun this section
library(data.table)
#write.csv(pd_final, "06-11-19 pd_final with epithelial duplicates.csv")
#write.csv(beta_final, "06-11-19 beta_final with epithelial duplicates.csv")
```

## now actually do the correlation test
```{r}
library(data.table)
#load datasets
pd_final_dups <- read.csv(file = "06-11-19 pd_final with epithelial duplicates.csv")
beta_final_dups <- read.csv(file = "06-11-19 beta_final with epithelial duplicates.csv")
pd.qc <- readRDS("pd.qc.rds")

#mean center the beta matrix
beta_rowmean <- rowMeans(beta_final_dups)
beta_mean_center <- (beta_final_dups - beta_rowmean)
beta_mean_center[1:5,1:5]

#subset the orig/duplicate IDs
dups <- pd.qc[pd.qc$NOTE == "technical duplicate", "Sample.ID"]
#take out the Dupl ending to get the original endings
orig <- gsub("Dupl", "", dups)
#subset original pd to get originals and duplicates
pd_dups <- pd.qc[pd.qc$Sample.ID %in% orig | pd.qc$Sample.ID %in% dups,]
#put the same samples next to each other
pd_dups <- pd_dups[order(pd_dups$Sample.ID), c("Sample.ID",
                                               "Male_label",
                                               "predicted_sex",
                                               "control_probe_flag",
                                               "mf_probe_fail_pct",
                                               "MQC",
                                               "UQC")]

library(tidyverse)
#using the pd_dups dataset which includes the mf_probe_fail_pct
#add sample ID as a column
sample1 <- gsub("_large", "", pd_dups$Sample.ID)
sample2 <- gsub("Dupl", "", sample1)
sample <- gsub("SAL", "", sample2)
sample
pd_dups$Sample <- sample
head(pd_dups)

#make pd datasets of the original samples and the duplicates
orig <- pd_dups[!duplicated(pd_dups$Sample), ]
dupl <- pd_dups[duplicated(pd_dups$Sample), ]

#merge in meth_id
orig$meth_id <- pd.qc[match(orig$Sample.ID, pd.qc$Sample.ID), 'meth_id']
dupl$meth_id <- pd.qc[match(dupl$Sample.ID, pd.qc$Sample.ID), 'meth_id']
colnames(orig)


#get the mean centered beta values for orig and dupl samples
orig_beta <- beta_mean_center[ , colnames(beta_mean_center) %in% orig$meth_id]
dim(orig_beta)
dim(beta_mean_center)
colnames(orig_beta)
dupl_beta <- beta_mean_center[,colnames(beta_mean_center) %in% dupl$meth_id]
dim(dupl_beta)
head(dupl_beta) #columns are the samples, rows are the probes

#run the correlation
cor <- cor(dupl_beta, orig_beta) #gives a 6x6 matrix of correlations

#rename the rownames from dupl_beta with the participant numbers
colnames(dupl_beta)
# "203287590025_R05C01" = 22
# "203287590025_R07C01" = 5
# "203287590048_R01C01" = 3
# "203287590228_R07C01" = 15
# "203287590228_R08C01" = 18
# "203293440009_R01C01" = 14
rownames(cor)[rownames(cor) == "203287590025_R05C01"] <- "22"
rownames(cor)[rownames(cor) == "203287590025_R07C01"] <- "5"
rownames(cor)[rownames(cor) == "203287590048_R01C01"] <- "3"
rownames(cor)[rownames(cor) == "203287590228_R07C01"] <- "15"
rownames(cor)[rownames(cor) == "203287590228_R08C01"] <- "18"
rownames(cor)[rownames(cor) == "203293440009_R01C01"] <- "14"

#rename the colnames from orig_beta with the participant numbers
colnames(orig_beta)
# "203282450208_R04C01" = 22
# "203282450208_R07C01" = 5
# "203286230053_R02C01" = 3
# "203286230103_R03C01" = 15
# "203287590048_R06C01" = 14
# "203287590228_R01C01" = 18
colnames(cor)[colnames(cor) == "203282450208_R04C01"] <- "22"
colnames(cor)[colnames(cor) == "203282450208_R07C01"] <- "5"
colnames(cor)[colnames(cor) == "203286230053_R02C01"] <- "3"
colnames(cor)[colnames(cor) == "203286230103_R03C01"] <- "15"
colnames(cor)[colnames(cor) == "203287590048_R06C01"] <- "14"
colnames(cor)[colnames(cor) == "203287590228_R01C01"] <- "18"

#r correlations:
    #note: 22 dupl was excluded by snp check so excluding it here
  # 22 = -0.6539773
  # 5 = 0.8811192
  # 3 = 0.9570237
  # 15 = 0.9188102
  # 14 = 0.8360845
  # 18 = 0.9449965
#average correlation

average <- (0.8811192+ 0.9570237+ 0.9188102+ 0.8360845+ 0.9449965)/ 5

average #0.9076068
```


# Drop cd45neg and sal_19_large from pd
## SAL_19_large mapped with immune cells according to EpiDISH (>80%)
```{r}
#drop from pd_dropdups
head(pd_dropdups) #cd45neg is in celltype
table(pd_dropdups$celltype)
# CD45neg CD45pos   large oragene   whole 
#       9      21      21       4      19 
cd45neg_samps <- pd_dropdups$celltype == "CD45neg"
table(cd45neg_samps)
# FALSE  TRUE 
#    65     9 
#total: 74
pd_final_saliva1 <-  pd_dropdups[!cd45neg_samps, ]
table(pd_final_saliva1$celltype)
# CD45pos   large oragene   whole 
#      21      21       4      19
#total: 65


#drop sal_19_large from pd
head(pd_final_saliva1)
pd_final_saliva2 <- pd_final_saliva1[pd_final_saliva1$Sample.ID != "SAL19_large", ]
table(pd_final_saliva2$celltype)
#CD45pos   large oragene   whole 
#     21      20       4      19 
#total: 64

#This is why sal_19_large was dropped from pd/beta
# library(ewastools)
# #the reference panel in estimateLC was created from ENCODE and Reinius data and dropped into the ewastools data file
# sorted19 <- estimateLC(beta_final, ref = "lauren_encode_reinius_ref")
# sorted19$name <- pd_final$Sample.ID
# #sal_19_large was 74% immune and only 23% epithelial


pd_final_saliva <- pd_final_saliva2
```



# 17_large doesn't agree with itself based on SNPs - see code file 2.0 saliva_preprocess
```{r}
#keep everything except 17_large;
pd_dropagree <- pd_final_saliva[pd_final_saliva$Sample.ID != "SAL17_large", ]
#check the dimensions;
dim(pd_dropagree) #63 samples 64 columns
```

# Drop low detection p samples: this is the >3%;
```{r}
#keep samples that have a fail rate of less than 0.03;
low_pdrop <- pd_dropagree[pd_dropagree$mf_probe_fail_pct <0.03, ]
#check the dimensions;
dim(low_pdrop) #dropped 3 more samples;
table(low_pdrop$celltype)
#total: 60 samples
```

# Drop probes;

# Drop probes that fail in >5% samples (detection p) - from final sample drop dataset - with whole and oragene;
```{r}
failedP <- detP > 0.01
failedP[1:5, 1:5] #false = not failed

#subset down to the 70 samples that were used (low_pdrop);
failedP <- failedP[ ,low_pdrop$meth_id] #meth_id works because colnames are the same as meth_id in detP and lowpdrop
#now the failed p probes is selected for only those in the 70 samples;

# recalculate per.probe
per.probe <- rowMeans(failedP) #finding the mean of each probe (1=true fail or 0) where 
table(per.probe >0.05)
# FALSE   TRUE 
# 855793  10298

#remove these two from memory because they no longer help
rm(detP, failedP)

#this has only the failed probes- if a probe fails in >5% it gets dropped
detp_fail <- per.probe[per.probe >0.05] 

#checking the number is the same;
length(detp_fail) #10298
head(detp_fail)

detp_fail <- names(detp_fail)

#starting with this many probes;
dim(noob_beta) #866091 

#keep anything that is not the failed row names;
beta_nodetp <- noob_beta[!rownames(noob_beta) %in% detp_fail, ]
dim(beta_nodetp) #855800     80
```

# Drop cross reactive probes;
```{r}
head(cross_probes) #Probe is an actual column here as opposed to row names;

#Number of cross reactive probes;
dim(cross_probes) #43254

#this drops all cross reactive probes;
beta_nocross <- beta_nodetp[!rownames(beta_nodetp) %in% cross_probes$Probe, ]
dim(beta_nocross) #812962        80
```

# Dropping X and Y chromosome probes;
```{r}
#make a dataset of only X and Y;
sex_probes <- Locations[Locations$chr == "chrX" | Locations$chr == "chrY", ] 
dim(sex_probes) #19681

#check first 5 rows and columns of noob_beta dataset;
noob_beta[1:5, 1:5] 

#keeping rownames that is not in x and y probes;
beta_noXY <- beta_nocross[!rownames(beta_nocross) %in% rownames(sex_probes), ]

#check dimensions of dataset without x and y probes;
dim(beta_noXY) #795694         80
```


# Match up the sample demographics and subsetted probes;
```{r}
#rename sample demographics dataset;
pd_final <- low_pdrop
dim(pd_final) #contains 60 samples including whole/oragene, does not include SAL_19_large

#meth_ids match column names, orders matrix;
beta_final <- beta_noXY[ , pd_final$meth_id] 
dim(beta_final) #795694     60

table(pd_final$celltype)

#save both pd_final and beta_final now because they still have whole and oragene in them
setwd("~/Research 2019")
saveRDS(beta_final, "06-11-19 beta_final_all_samps.rds")
saveRDS(pd_final, "06-11-19 pd_final_all_samps.rds")
```


# Dropping whole and oragene samples;
```{r}
pd_final <- pd_final[!pd_final$celltype %in% c("whole", "oragene"), ]
dim(pd_final) #39 fractions 64 variables

#picking only columns that are in pd_final and keeping them;
beta_final <- beta_final[ , pd_final$meth_id]
dim(beta_final) #38 samples that do not include whole and oragene or prev dropped samples;
```

# Save final files;
```{r}
#this includes the subsetted sample data (demographic data), no whole or oragene or cd45neg
saveRDS(pd_final, file = "06-11-19 pd_final.rds")

#save the beta matrix - no whole, oragene, or cd45neg;
saveRDS(beta_final, file = "06-11-19 beta_final.rds")
```

# Double check files
```{r}
pd_final <- readRDS("06-11-19 pd_final.rds") #38 samples, 64 variables
beta_final <- readRDS("06-11-19 beta_final.rds")
dim(beta_final) #795694 probes
```