# Saliva cell type DNA methylation reference panel for epidemiological studies in children

## Citation Information
Middleton LYM, Dou JF, Fisher J, Heiss JA, Nguyen V, Just AC, Faul JD, Ware EB, Mitchell C, Colacino JA, Bakulski KM. 2021. Saliva cell type DNA methylation reference panel for epidemiology studies in children. Epigenetics. PMID: 33588693, PMCID: PMC8865319.  DOI: 10.1080/15592294.2021.1890874

This Github repository contains the data management and analytic scripts to produce the following manuscript:[Saliva cell type DNA methylation reference panel for epidemiological studies in children](https://pubmed.ncbi.nlm.nih.gov/33588693/)

## Abstract
Saliva is a widely used biological sample, especially in pediatric research, containing a heterogenous mixture of immune and epithelial cells. Associations of exposure or disease with saliva DNA methylation can be influenced by cell-type proportions. Here, we developed a saliva cell-type DNA methylation reference panel to estimate interindividual cell-type heterogeneity in whole saliva studies. Saliva was collected from 22 children (7-16 years) and sorted into immune and epithelial cells, using size exclusion filtration and magnetic bead sorting. DNA methylation was measured using the Illumina MethylationEPIC BeadChip. We assessed cell-type differences in DNA methylation profiles and tested for enriched biological pathways. Immune and epithelial cells differed at 181,577 (22.8%) DNA methylation sites (t-test p < 6.28 × 10-8). Immune cell hypomethylated sites are mapped to genes enriched for immune pathways (p < 3.2 × 10-5). Epithelial cell hypomethylated sites were enriched for cornification (p = 5.2 × 10-4), a key process for hard palette formation. Saliva immune and epithelial cells have distinct DNA methylation profiles which can drive whole-saliva DNA methylation measures. A primary saliva DNA methylation reference panel, easily implemented with an R package, will allow estimates of cell proportions from whole saliva samples and improve epigenetic epidemiology studies by accounting for measurement heterogeneity by cell-type proportions.

## Funding
We thank the University of Michigan Epigenomics Core and Advanced Genomics Core for DNA methylation measures. Support for this research was provided by the National Institute of Environmental Health Sciences (grants P30 ES017885, R01 ES028802, U01 ES026697, R35 ES031686, R01 ES025531, R01 ES025574), the National Institute on Aging (grants R01 AG067592, R01 AG060110-01), the National Institute on Minority Health and Health Disparities (grants R01 MD011716, R01 MD013299), and the National Institutes of Health Office of the Director (grants UG3 OD023285, UH3 OD023285).


